airflow/providers/google/cloud/secrets/secrets_manager.py

```diff
@@ -0,0 +1,132 @@
+from airflow.providers.google.cloud.utils.credentials_provider import (
+    _get_scopes, get_credentials_and_project_id,
+)
+from airflow.secrets import BaseSecretsBackend
+from airflow.utils.log.logging_mixin import LoggingMixin
+
+
+class CloudSecretsManagerSecretsBackend(BaseSecretsBackend, LoggingMixin):
+    """
+    Retrieves Connection object from GCP Secrets Manager
+
+    Configurable via ``airflow.cfg`` as follows:
+
+    .. code-block:: ini
+
+        [secrets]
+        backend = airflow.providers.google.cloud.secrets.secrets_manager.CloudSecretsManagerSecretsBackend
+        backend_kwargs = {"connections_prefix": "airflow/connections"}
+
+    For example, if secret id is ``airflow/connections/smtp_default``, this would be accessible
+    if you provide ``{"connections_prefix": "airflow/connections"}`` and request conn_id ``smtp_default``.
+
+    :param connections_prefix: Specifies the prefix of the secret to read to get Connections.
+    :type connections_prefix: str
+    :param gcp_key_path: Path to GCP Credential JSON file;
+        use default credentials in the current environment if not provided.
+    :type gcp_key_path: str
+    :param gcp_scopes: Comma-separated string containing GCP scopes
+    :type gcp_scopes: str
+    """
+    def __init__(
+        self,
+        connections_prefix: str = "airflow/connections",
+        gcp_key_path: Optional[str] = None,
+        gcp_scopes: Optional[str] = None,
+        **kwargs
+    ):
+        self.connections_prefix = connections_prefix.rstrip("/")
+        self.gcp_key_path = gcp_key_path
+        self.gcp_scopes = gcp_scopes
+        self.credentials: Optional[str] = None
+        self.project_id: Optional[str] = None
+        super().__init__(**kwargs)
+
+    @cached_property
+    def client(self) -> SecretManagerServiceClient:
+        """
+        Create an authenticated KMS client
+        """
+        scopes = _get_scopes(self.gcp_scopes)
+        self.credentials, self.project_id = get_credentials_and_project_id(
+            key_path=self.gcp_key_path,
+            scopes=scopes
+        )
+        _client = SecretManagerServiceClient(
+            credentials=self.credentials,
+            client_info=ClientInfo(client_library_version='airflow_v' + version.version)
+        )
+        return _client
+
+    def build_secret_id(self, conn_id: str) -> str:
+        """
+        Given conn_id, build path for Secrets Manager
+
+        :param conn_id: connection id
+        :type conn_id: str
+        """
+        secret_id = f"{self.connections_prefix}/{conn_id}"
+        return secret_id
+
+    def get_conn_uri(self, conn_id: str) -> Optional[str]:
+        """
+        Get secret value from Secrets Manager.
+
+        :param conn_id: connection id
+        :type conn_id: str
+        """
+        secret_id = self.build_secret_id(conn_id=conn_id)
+        # always return the latest version of the secret
+        secret_version = "latest"
```

Comment:
Hey there. I'm Seth from the Secret Manager team. Thanks for building this integration. Feel free to reach out if you have future questions.

We strongly discourage users from pinning to "latest" in production workloads. Users should specifically pin to a version (e.g. 1, 2, 3). Is there a way we can update this to allow the user to specify a version id?

Reply:
Hello. I'm Kamil from Polidea. We are a Google partner and we are building many integrations for Google services at the request of [Cloud Composer](https://cloud.google.com/composer).

We have an abstraction layer that does not provide for this possibility, but I think we can introduce. special parameter syntax that lets you reference a specific version. 

But I will explain the small context of how this change works so that you can make better design decisions. In previous versions of Airflow, all data was stored in a database (this behavior may have been overwritten by environment variables, so you can also store secret data in env var).

The user selects the connection using the operator parameter. The following example will use connections with ID = `'my-company-id``.
```
    download_file = GCSToLocalOperator(
        task_id="download_file",
        object_name=BUCKET_FILE_LOCATION,
        bucket=BUCKET_1,
        filename=PATH_TO_SAVED_FILE,
        gcp_conn_id='my-company-id`
    )
```
Please note that the user does not specify the type of backend secret. It is a universal mechanism and this data may come from a database, from environment variables or from another service.

What do you think so that the user can define the connection version after the colon? The sample call will then look like this.
```
    download_file = GCSToLocalOperator(
        task_id="download_file",
        object_name=BUCKET_FILE_LOCATION,
        bucket=BUCKET_1,
        filename=PATH_TO_SAVED_FILE,
        gcp_conn_id='my-company-id:2`
    )
```
It won't be a perfect solution, but I think it's easy to implement and easy to use.

Reply:
@sethvargo Thanks for letting us know. Why is it not recommended, at least from Airflow's perspective, our users would want to pull the secrets with the most recent version (hence 'latest').

Can you give us some context to let us know how a user who does not know how many versions of a secret exists pull it?

Reply:
Hey @kaxil - [here's the protocol](https://github.com/spring-cloud/spring-cloud-gcp/pull/2302) we'd like most integrators to try and align on.

Latest is a floating alias that resolves to "the highest numbered secret version" for the secret. Often times, the team creating the secret is separate from the team consuming the secret. In these cases, the latest version may not be active or usable yet.

Generally, you want a workload to be coupled to a specific secret so you can roll forward and roll backward in the event something breaks.

It's fine to allow users to _specify_ latest, but I don't think it should be the default behavior. You can see how we recommend specifying the format in the linked issue above.